File: ICPC/SegundaFecha2023/b.py
import numpy as np
import functools
MOD = 1000000007
n = 1
k = 1
maxN = 13
arre = np.empty(maxN, np.int32)
fact = np.empty(maxN, np.int64)
invFacto = np.empty(maxN, np.int64)
bucket = {}

def binPow(base, power):
    resp = 1
    multi = base
    while power > 0:
        if power & 1:
            resp = (resp*multi)%MOD
        power >>= 1
        multi = (multi*multi)%MOD
    return resp

def precalcula():
    fact[0] = 1
    fact[1] = 1
    for i in range(2, maxN):
        fact[i] = (fact[i-1] * i)%MOD
    global invFacto
    invFacto = np.fromiter(map(lambda x: binPow(x, MOD-2), fact), dtype=np.int32)
   
 
n, k = map(int, input().split(' '))
arre = np.fromiter(map(int, input().split(' ')), np.int32)
precalcula()
for x in arre:
    if bucket.get(x) == None:
        bucket[x] = 1
    else:
        bucket[x] += 1
resp = functools.reduce(lambda a, b: (a*fact[b])%MOD, bucket.values()) 
resp = (resp*invFacto[n])%MOD

print(resp)
for _ in range(k):
    a,b = map(int, input().split(' '))
    a-=1
    
    resp = (resp*invFacto[bucket[arre[a]]])%MOD
    bucket[arre[a]] -= 1
    resp = (resp*fact[bucket[arre[a]]])%MOD

    arre[a] = b
    if bucket.get(b) == None:
        bucket[b] = 1
    else:
        bucket[b] += 1
    resp = (resp*bucket[b])%MOD
    print(resp)

File: ICPC/SegundaFecha2023/b.cpp
#include <algorithm>
#include <iostream>
#include <unordered_map>

using namespace std;
typedef int64_t ll;
#define MOD 1000000007
int n, k;
int const maxN = 1e6+1;
int arre[maxN];
ll fact[maxN], invFact[maxN];
unordered_map<int, int> bucket;

ll binPow(ll base, ll power){
    ll resp = 1, multi = base;
    while(power > 0){
        if(power & 1){
            resp = (resp*multi)%MOD;
        }
        power >>= 1;
        multi = (multi*multi)%MOD;
    }
    return resp;
}

void precalcula(){
    fact[0] = fact[1] = invFact[1] = 1;
    for(int i = 2; i < maxN; ++i){
        fact[i] = (fact[i-1] * i)%MOD;
        invFact[i] = binPow(fact[i], MOD-2);
    }
}
 
int main(){
    ios_base::sync_with_stdio(0); cin.tie(0);
    cin>>n>>k;
    for(int i =0; i < n; ++i){
        cin>>arre[i];
    }
    precalcula();
    
    for(int i = 0; i < n; ++i){
        ++bucket[arre[i]];
    }

    ll resp = invFact[n];
    for( pair<int, int> keyValue : bucket){
        resp = (resp*fact[keyValue.second])%MOD;       
    }
    cout<<resp<<"\n";

    int a, b;
    for(int i = 0; i < k; ++i){
        cin>>a>>b;
        --a;
        resp = (resp*invFact[bucket[arre[a]]])%MOD;
        --bucket[arre[a]];
        resp = (resp*fact[bucket[arre[a]]])%MOD;

        arre[a] = b;
        ++bucket[b];
        resp = (resp*bucket[b])%MOD;
        cout<<resp<<"\n";
    }
    return 0;
}